README
NAME
    Devel::Cover - Code coverage metrics for Perl

SYNOPSIS
    To get coverage for an uninstalled module:

     cover -test

    or

     cover -delete
     HARNESS_PERL_SWITCHES=-MDevel::Cover make test
     cover

    To get coverage for an uninstalled module which uses Module::Build (0.26
    or later):

     ./Build testcover

    If the module does not use the t/*.t framework:

     PERL5OPT=-MDevel::Cover make test

    If you want to get coverage for a program:

     perl -MDevel::Cover yourprog args
     cover

     perl -MDevel::Cover=-db,cover_db,-coverage,statement,time yourprog args

DESCRIPTION
    This module provides code coverage metrics for Perl. Code coverage
    metrics describe how thoroughly tests exercise code. By using
    Devel::Cover you can discover areas of code not exercised by your tests
    and determine which tests to create to increase coverage. Code coverage
    can be considered as an indirect measure of quality.

    I consider this software to have an alpha status.  By that I mean that I
    reserve the right to alter the interface in a backwards incompatible
    manner without incrementing the major version number.  I specifically do
    not mean that this software is full of bugs or missing key features. 
    Although I'm making no guarantees on that front either.  In short, if
    you are looking for code coverage software for Perl, you have probably
    come to the end of your search.  For more of my opinions on this
    subject, see http://pjcj.sytes.net/notes/2007/03/14#alpha

    Code coverage data are collected using a pluggable runops function which
    counts how many times each op is executed.  These data are then mapped
    back to reality using the B compiler modules.  There is also a statement
    profiling facility which needs a better backend to be really useful. 
    This release also includes an experimental mode which replaces ops
    instead of using a pluggable runops function.  This provides a nice
    speed increase, but needs better testing before it becomes the default. 
    You probably don't care about any of this.

    The cover program can be used to generate coverage reports.

    Statement, branch, condition, subroutine, pod and time coverage
    information is reported.  Statement coverage data should be reasonable,
    although there may be some statements which are not reported.  Branch
    and condition coverage data should be mostly accurate too, although not
    always what one might initially expect.  Subroutine coverage should be
    as accurate as statement coverage.  Pod coverage comes from
    Pod::Coverage.  If Pod::Coverage::CountParents is available it will be
    used instead.  Coverage data for path coverage are not yet collected.

    The gcov2perl program can be used to convert gcov files to
    "Devel::Cover" databases.

    You may find that the results don't match your expectations.  I would
    imagine that at least one of them is wrong.

    The most appropriate mailing list on which to discuss this module would
    be perl-qa.  Discussion has migrated there from perl-qa-metrics which is
    now defunct.  See <http://lists.perl.org/list/perl-qa.html>.

    The Devel::Cover repository can be found at
    <http://github.com/pjcj/Devel--Cover>.

REQUIREMENTS
    *   Perl 5.6.1 or greater.  Perl 5.8.8 or greater is recommended.

        Perl 5.7.0 is unsupported.  Perl 5.8.8 or greater is recommended. 
        Perl 5.8.7 has problems and may crash.  Whilst Perl 5.6 should
        mostly work you will probably miss out on coverage information which
        would be available using a more modern version and will likely run
        into bugs in perl.  Perl 5.8.0 will give slightly different results
        to more recent versions due to changes in the op tree.

    *   The ability to compile XS extensions.

        This means a working compiler and make program at least.

    *   Storable and Digest::MD5

        Both are in the core in Perl 5.8.0 and above.

    *   Template and PPI::HTML or Perl::Tidy

        if you want syntax highlighted HTML reports.

    *   Pod::Coverage

        if you want Pod coverage.

    *   Test::More

        in order to run the tests

    *   Test::Warn

        in order to run some of the tests

    *   Test::Differences

        if the tests fail and you would like nice output telling you why.



CHANGES
Devel::Cover.pm history

Release 0.78 - 17th May 2011
 - Fix up test quoting to work with Windows too.

Release 0.77 - 15th May 2011
 - Fix cover -test covering all the test files (Larry Leszczynski) (rt 65920).
 - Add DEVEL_COVER_IO_OPTIONS environment variable.
 - Sort runs by start time.
 - Add digests to DB.
 - Add cover -make option (Olivier Mengué) (rt 44906).
 - Add digests to DB.
   This should fix some problems related to losing coverage data when there are
   duplicate files.  This happens most usually when modules are sometimes loaded
   from lib and sometimes from blib. (rt 14192, 32465, 45737).
 - Add branch coverage for gcov (rt 30365).

Release 0.76 - 18th April 2011
 - Move CHANGES file into root for search.cpan.org (rt 67541).
 - Add top level version subroutine (requested by H.Merijn Brand).
 - Add DEVEL_COVER_DB_FORMAT environment variable.
 - Add advisory locking to database IO operations.
 - Explain what to do if regexp_eval.t ever fails again (Florian Ragwitz).
 - Remove leftovers from PERL_OBJECT (Florian Ragwitz).
 - Recommend 5.8.8 and above - 5.8.7 is crashing so skip some tests there.

Release 0.75 - 17th April 2011
 - Write database as JSON if JSON::PP is available.

Release 0.74 - 16th April 2011
 - Test against 5.12.3 and code frozen 5.14.0 (unreleased).
 - Fix tests to work with all releases of 5.13.x (rt 64210, 60901).
 - Avoid race condition writing DB structure files (Nicholas Clark).
 - Add debuglog method to Devel::Cover::DB::Structure (Nicholas Clark).
 - Be more careful deleting DB structure files (Nicholas Clark).
 - Get cover -test to honour the db passed in.

Release 0.73 - 2nd October 2010
 - Tidy up Makefile.PL and META.yml.
 - Get coverage working with Windows again.

Release 0.72 - 27th September 2010
 - Teach cpancover how to work in parallel.
 - Improve speed of check_file() (Goro Fuji).
 - Fix errors with non-existent directories on Cygwin (Goro Fuji).
 - Improve self coverage with new test.
 - Fix tests to work with 5.13.5.
 - New dependency on Test::Warn.
 - New dependency on Parallel::Iterator for cpancover.

Release 0.71 - 10th September 2010
 - Improve running of Devel::Cover on itself.
 - Fix occasional parallel test failures.
 - Test against 5.12.2.

Release 0.70 - 29th August 2010
 - Get Devel::Cover working better on itself.
 - Distribution should now pass on 5.13.* development releases.

Release 0.69 - 28th August 2010
 - Correctly report on C<my $x = $y || return> and other shortcuts.
 - Put end to end tests in t/e2e.
 - Add test for regexp eval fail (Florian Ragwitz).
 - Fix some warnings from strict compilers (Florian Ragwitz).
 - Allow tests to run in parallel (Florian Ragwitz).
 - Test against 5.13.* development releases.
 - We now require Test::More to run the tests.

Release 0.68 - 5th August 2010
 - Fix gcov2perl to work with large numbers (Thomas Dorner) (rt 45028).
 - Fix "gcov -l" include files (Thomas Dorner) (rt 44864).
 - Test against 5.12.1.
 - Gross workaround for for regexp evals (Florian Ragwitz).

Release 0.67 - 8th May 2010
 - Fix up the Makefile for dmake on Strawberry Perl (Curtis Jewell) (rt 50774).

Release 0.66 - 12th April 2010
 - Move to faster method of collecting data.  Could be up to twice as fast now.
 - Add -replace_ops options to be able to revert to previous collection method.
 - Test against 5.12.0 and update tests as necessary.
 - Work better with non-existent databases.
 - XS fixes related to overriding ops rather than replacing runops
   (Florian Ragwitz).
 - Don't chmod created directories (mkdir should suffice).

Release 0.65 - 8th August 2009
 - Fix uninitialised value warning (reported by Andrew Billeb).
 - Test against 5.8.9 and 5.10.1 (RC1).
 - Correct deparsed conditional output in elsif conditions.
 - Unset PERL5OPT when running tests.

Release 0.64 - 10th April 2008
 - Build on Windows.
 - Make "cover -test" work for Module::Build (Michael G Schwern) (rt 34263).
 - Make "cover -test" cover .[ch] files (Michael G Schwern) (rt 34262).
 - Make "cover -gcov" a valid option (Michael G Schwern) (rt 34261).

Release 0.63 - 16th November 2007
 - require 5.006001 might stop CPAN testers trying to test with 5.005.
 - store_return() should be a void function (H.Merijn Brand).
 - Finish dor support.
 - Add support for exec (Brandon Black).
 - Tested against 5.10 (RC1).

Release 0.62 - 5th November 2007
 - Add table sorting to Html_basic report (Nathan Haigh).
 - Assume heredocs are constants.
 - Don't add 0x200 to $^P ("file" names for evals) (Fix for Template::Declare).
 - Add timer and alarm for cpancover.
 - Add Report Date to basic html report.
 - Quieten some warnings.
 - Document cover -test.
 - Get rid of any __DIE__and __WARN__ signals during the report.
 - Don't try to get a digest for "-e".
 - Tidy up filename normalisation.
 - Be more explicit about what I mean when I call this alpha software.
 - Add uncoverable comments.
 - Don't complain about POSIX.pm (Erwan Lemonnier).
 - Don't resolve pathnames of symbolic links (Stefan Becker).
 - Spelling nits (James E Keenen).

Release 0.61 - 10th January 2007
 - Fix some "ignored" errors due to multiple subs of the same name on the same
   line (rt 14192).
 - Display pod coverage with subroutine coverage in text report.
 - Update golden output (tests were failing in 0.60).

Release 0.60 - 2nd January 2007
 - Simplify get_key function and remove TODO item (Gisle Aas).
 - Be careful with UNC paths on Windows (John LoVerso) (rt 24105).
 - Only call HvSHAREKEYS_off if threading is enabled.
 - s/unvailable/unavailable/ (Jim Cromie).
 - Don't key on op_targ - it might change (rt 22701).
 - Tidy up docs and error messages for reports (rt 21098).
 - Better database validation.
 - Don't delete invalid datbases (rt 16039).
 - Tested against 5.9.5.
 - Allow for new anonlist and anonhash ops (rt 24067).
 - Use outputfile for both cover and cpancover.
 - Add compilation report (Denis Howe).

Release 0.59 - 23rd August 2006

Makefile.PL
#!/usr/bin/perl

# Copyright 2001-2011, Paul Johnson (pjcj@cpan.org)

# This software is free.  It is licensed under the same terms as Perl itself.

# The latest version of this software should be available from my homepage:
# http://www.pjcj.net

require 5.006001;

use strict;
use warnings;

use Cwd;
use ExtUtils::MakeMaker;
use ExtUtils::Manifest "maniread";
use File::Copy;

$| = 1;

my $Version  = "0.78";
my $Date     = "17th May 2011";
my $Author   = 'pjcj@cpan.org';

my @perlbug  = ("perlbug", "-a", $Author,
                           "-s", "Installation of Devel::Cover $Version");
my $Perlbug  = join " ", map { / / ? "'$_'" : $_ } @perlbug;

my @files    = sort keys %{maniread()}, "lib/Devel/Cover/Inc.pm";
my @versions = grep !/^(?:README|Makefile\.PL|\.gitignore)$/, @files;

my $base     = getcwd;

my %inc      = map { -d $_ ? (($_ eq "." ? $_ : Cwd::abs_path($_)) => 1) : () }
                   @INC;
my @inc      = sort keys %inc;

open I, ">lib/Devel/Cover/Inc.pm"
    or die "Cannot open lib/Devel/Cover/Inc.pm: $!";
print I <<"EOI";
# Copyright 2001-2011, Paul Johnson (pjcj\@cpan.org)

# This software is free.  It is licensed under the same terms as Perl itself.

# The latest version of this software should be available from my homepage:
# http://www.pjcj.net

# This file was automatically generated by Makefile.PL.

package Devel::Cover::Inc;

use strict;
use warnings;

our \$VERSION      = "$Version";
our \$Perl_version = '$]';
our \$Base         = '$base';
our \@Inc          = qw( @inc );
chomp (our \$Perl  = <<'EOV');  # Careful with \\\\ in the path
$^X
EOV

if (\$Perl_version ne \$])
{
    print STDERR <<"EOM";

This version of Devel::Cover was built with Perl version \$Perl_version.
It is now being run with Perl version \$].
Attempting to make adjustments, but you may find that some of your modules do
not have coverage data collected.  You may need to alter the +-inc, +-ignore
and +-select options.

EOM
    eval "use Cwd";
    my \%inc = map { -d \$_ ? ((\$_ eq "." ? \$_ : Cwd::abs_path(\$_)) => 1) : () }
                   \@INC;
    \@Inc = sort keys \%inc;
}

# TODO - check for threadedness, 64bits etc. ?

1
EOI
close I or die "Cannot close lib/Devel/Cover/Inc.pm: $!";

print "Writing tests ........ ";

for my $d (qw( t t/e2e ))
{
    unless (-d $d)
    {
        mkdir $d or die "Cannot mkdir $: $!";
    }
}

my @tests;
opendir D, "tests" or die "Cannot opendir tests: $!";
for my $t (readdir D)
{
    next unless -f "tests/$t";
    next if $t =~ /\.(pm|pl|org|bak|uncoverable)$/;
    next if $t =~ /~$/;

    push @tests, $t;

    if ($t =~ /\.t/)
    {
        copy("tests/$t", "t/e2e/$t") or die "Cannot copy tests/$t to t/e2e/$t: $!";
        next
    }

    open T, ">t/e2e/a$t.t" or die "Cannot open t/e2e/a$t.t: $!";
    print T <<EOT;
#!$^X

# Copyright 2002-2011, Paul Johnson (pjcj\@cpan.org)

# This software is free.  It is licensed under the same terms as Perl itself.

# The latest version of this software should be available from my homepage:
# http://www.pjcj.net

use strict;
use warnings;

use lib "$base/lib";
use lib "$base/blib/lib";
use lib "$base/blib/arch";
use lib "$base/t";

use Devel::Cover::Test $Version;

Devel::Cover::Test->new("$t");
EOT
    close T or die "Cannot close t/e2e/a$t.t: $!";
}
closedir D or die "Cannot closedir tests: $!";

s/^/tests\// for @tests;
push @tests, grep !/e2e/, glob "t/*/*.t";

print "done\n\n";

sub check
{
    my ($module, $text, $version) = @_;

    printf "checking for %-18s %-16s .... ",
           $module, $version ? "version $version" : "";

    eval "use $module";
    (my $mod = $module) =~ s/::/\//g;
    if (my $m = $INC{"$mod.pm"})
    {
        my $v = eval { no warnings; eval "\$${module}::VERSION" };
        printf "%-8s $m\n", $v;
        if ($version && $v < $version)
        {
            print "\n\n\n$text\n\n";
        }
    }
    else
    {
        print "not found\n\n\n$text\n\n";
    }
};

check "Storable", <<EOM;
Storable is required to store the coverage database.  You can download
Storable from CPAN.
EOM

check "JSON::PP", <<EOM;
JSON::PP is required to store the coverage database in a portable format.
This is necessary in order to merge coverage data from incompatible systems.
Storing the database in a portable format will not be available until you
install JSON::PP.  In the meantime you may continue to use the rest of
Devel::Cover.
EOM

check "Digest::MD5", <<EOM;
Digest::MD5 is required to check whether covered files have changed.  You can
download Digest::MD5 from CPAN.
EOM

check "Test::More", <<EOM;
Test::More is required to run the Devel::Cover tests.  You can download
Test::More from CPAN.
EOM

check "Test::Warn", <<EOM;
Test::Warn is required to run some of the Devel::Cover tests.  You can
download Test::Warn from CPAN.
EOM

check "Template", <<EOM, "2.00";
Template 2.00 is required to run the some HTML backends to cover and
for cpancover.  Unless you have specific requirements this should not be
a problem, but you will not be able to use these reports until you
install the Template Toolkit, available from CPAN.  In the meantime you
may continue to use the rest of Devel::Cover.
EOM

my $m = <<EOM;
One of PPI::HTML 1.07 or Perl::Tidy 20060719 is required to add syntax
highlighting to some HTML backends to cover and for cpancover.  Unless
you have specific requirements this should not be a problem, but you will
not be able to use syntax highlighting in these reports until you install
PPI::HTML or Perl::Tidy, available from the CPAN.  In the meantime you
may continue to use the rest of Devel::Cover.
EOM

check "PPI::HTML", $m, "1.07";

check "Perl::Tidy", $m, "20060719";

print "checking for Pod::Coverage      version 0.06     .... ";
my $e = <<EOM;
Pod::Coverage 0.06 is required to do pod coverage.  This will tell
you how well you have documented your modules.  Pod coverage will be
unavailable until you install this module, available from CPAN.  In the
meantime, you may continue to use the rest of Devel::Cover.
EOM

eval "use Pod::Coverage";
if (my $m = $INC{"Pod/Coverage.pm"})
{
    my $v = eval { no warnings; $Pod::Coverage::VERSION };
    print $v < 0.06 ? "$v\n\n\n$e\n\n" : "$v     $m\n";
    print "checking for Pod::Coverage::CountParents         .... ";
    $e = <<EOM;
    Pod::Coverage::CountParents.pm is used for Pod coverage if it is available.
    We will fall back to using Pod::Coverage.pm.  If you want to use
    Pod::Coverage::CountParents.pm, just install it from CPAN.
EOM
    eval "use Pod::Coverage::CountParents";
    if (my $m = $INC{"Pod/Coverage/CountParents.pm"})
    {
        my $v = eval { no warnings; $Pod::Coverage::CountParents::VERSION };
        $v ||= "    ";
        print "$v     $m\n";
    }
    else
    {
        print "not found\n\n\n$e\n\n";
    }
}
else
{
    print "not found\n\n$e\n";
}

check "Test::Differences", <<EOM;
Test::Differences is used to display output from failed tests.
Hopefully there won't be any failed tests, but if there are you will get
output that may not be a model of clarity.  If you do get test failures
and you fancy helping me by debugging them, then you might like to
consider installing Test::Differences.  You can download
Test::Differences from CPAN.
EOM

my $latest_tested = "5.014000";
print <<EOM if $] > $latest_tested;

Devel::Cover $Version has not been tested with perl $].
Testing will take place against expected output from perl $latest_tested.
You may well find failing tests.

EOM

if ($] < 5.008)
{
    print <<EOM;

Devel::Cover $Version is not fully functional on perl $].  It should
mostly work, but there are some constructs for which coverage will not
be collected, and you may well encounter bugs which have been fixed in
subsequent versions of perl.  Perl versions 5.8.8 and above should work
better.

EOM

    print <<EOM if $^O eq "MSWin32";
And things are even worse under Windows.  You may well find random bugs
of various severities.

EOM
}
elsif ($] < 5.008002)
{
    print <<EOM;

Devel::Cover $Version mostly works on perl $], but you may encounter
strange behaviours (bugs) which have been fixed in subsequent versions
of perl.  This is particularly true if you are running a threaded perl.
Perl versions 5.8.2 and above should work better.

EOM
}

print "\n" x 3;

$ExtUtils::MakeMaker::Verbose = 0;

WriteMakefile
(
    NAME             => "Devel::Cover",
    VERSION          => $Version,
    AUTHOR           => 'Paul Johnson (pjcj@cpan.org)',
    ABSTRACT_FROM    => "lib/Devel/Cover.pm",
    DIR              => [],
    EXE_FILES        => [ qw( cover gcov2perl cpancover ) ],
    PERL_MALLOC_OK   => 1,
    PREREQ_PM        => {
                            Storable      => 0,
                            "Digest::MD5" => 0,
                            "Test::More"  => 0,
                            "Test::Warn"  => 0,
                        },
    TYPEMAPS         => [ "utils/typemap" ],
    clean            => { FILES    => "t/e2e/* cover_db* t/e2e/*cover_db"  },
    depend           => { distdir  => "@files"                             },
    dist             => { COMPRESS => "gzip --best --force"                },
    test             => { TESTS    => "t/*/*.t"                            },
    realclean        => $] < 5.008008 ?
                        {
                            FILES => "lib/Devel/Cover/Inc.pm",
                            POSTOP => "\$(RM_RF) cover_db t/e2e"
                        } :
                        { FILES => "lib/Devel/Cover/Inc.pm cover_db t/e2e" },
    $] >= 5.010001 ?
    (
        LICENSE          => "perl",
        META_ADD         =>
        {
            requires =>
            {
                perl                          => 5.006001,
                Storable                      => 0,
                "Digest::MD5"                 => 0,
            },
            dynamic_config => 1,
        },
        META_MERGE       =>
        {
            recommends =>
            {
                perl                          => 5.008002,
                Template                      => 2.00,
                "PPI::HTML"                   => 1.07,
                "Perl::Tidy"                  => 20060719,
                "Pod::Coverage"               => 0.06,
                "Pod::Coverage::CountParents" => 0,
                "Test::Differences"           => 0,
                "Parallel::Iterator"          => 0,
                "JSON::PP"                    => 0,
            },
            build_requires =>
            {
                "Test::More"                  => 0,
                "Test::Warn"                  => 0,
            },
            no_index =>
            {
                directory => [ "tests", "utils" ],
            },
            resources =>
            {
              license     => "http://dev.perl.org/licenses",
              homepage    => "http://www.pjcj.net/perl.html",
              bugtracker  => "http://rt.cpan.org/NoAuth/Bugs.html?Dist=Devel-Cover",
              repository  => "http://github.com/pjcj/Devel--Cover",
              MailingList => "http://lists.perl.org/list/perl-qa.html",
            },
        },
        MIN_PERL_VERSION => "5.6.1",
    ) :
    ()
);

print "\n";
print <<EOM if 0;

--------------------------------------------------------------------------------

I like to have some idea of the distribution of this software.  To that end I
would be very grateful for mail from you.  This will be used only to satisfy my
curiosity and to help me make decisions which may involve changes to the
module.

If you can run perlbug you can send me a success report with "make ok".
Failure reports with "make nok" are also appreciated.

If you have any questions or comments, mailing list details are available in the
README file, or send mail to me directly.

EOM

sub MY::libscan
{
    my ($self, $path) = @_;
    (my $p = $path) =~ s/^\$\(INST_LIB\)/lib/;  # 5.6.1
    $p =~ s|\\|/|g if $^O eq "MSWin32";
    # print "$path $p\n";
    my $wanted = -d $p;                         # 5.9.0
    for my $f (@files)
    {
        # print "$p - $f\n";
        last if $wanted ||= $p =~ /$f$/;
    }
    $wanted && $path;
}

sub MY::postamble
{
    my %tests;
    @tests{@tests} = map { (my $t = $_) =~ s/\W/_/g; "cover_db_$t" } @tests;
    my @reports =
      qw(compilation html_basic html_minimal html html_subtle sort text2 text);

    qq[
SET_VERSION = \$(PERL) -pi \\
    -e 's/(^\\s*(?:our\\s+)\\\$\$VERSION = ")\\d+\\.\\d+(";)/\$\${1}$Version\$\$2/;' \\
    -e 's/(Version )\\d+\\.\\d+( - ).*/\$\${1}$Version\$\${2}$Date/;' \\
    -e 's/(Devel::Cover Version )\\d+\\.\\d+/\$\${1}$Version/;' \\
    -e 's/(\\buse Devel::Cover(?:::\\w+)*\\s+)\\d+\\.\\d+/\$\${1}$Version/;'

tags : @files
\t ptags @files

@versions : Makefile.PL
\t \$(SET_VERSION) @versions

README : lib/Devel/Cover.pm
\t TERMCAP= COLUMNS=80 pod2text -s lib/Devel/Cover.pm | \\
    \$(PERL) -n \\
    -e 'print if (/NAME/         ... /^[A-Z ]+\$\$/) =~ /^\\d+\$\$/;' \\
    -e 'print if (/SYNOPSIS/     ... /^[A-Z ]+\$\$/) =~ /^\\d+\$\$/;' \\
    -e 'print if (/DESCRIPTION/  ... /^[A-Z ]+\$\$/) =~ /^\\d+\$\$/;' \\
    -e 'print if (/REQUIREMENTS/ ... /^[A-Z ]+\$\$/) =~ /^\\d+\$\$/;' \\
    > README

ppm : ppd pure_all
\t tar cf Devel-Cover.tar blib
\t gzip --best --force Devel-Cover.tar
\t \$(PERL) -pi.bak \\
    -e 's/(OS NAME=")[^"]*/\$\$1MSWin32/;' \\
    -e 's/(ARCHITECTURE NAME=")[^"]*/\$\$1MSWin32-x86-multi-thread/;' \\
    -e 's/(CODEBASE HREF=")[^"]*/\$\$1Devel-Cover.tar.gz/;' \\
    Devel-Cover.ppd

TAINT = -T
COVER_OPTIONS =

_run : pure_all
\t \$(PERL) \$(TAINT) -Iblib/lib -Iblib/arch -MDevel::Cover=-merge,0,`\$(PERL) -e '\$\$l = qx|grep __COVER__ \$\$ARGV[0]|; \$\$l =~ /__COVER__\\s+criteria\\s+(.*)/; (\$\$c = \$\$1 || "all") =~ s/\\s+/,/g; \$\$p = "\$\$1," if \$\$l =~ /__COVER__\\s+test_parameters\\s+(.*)/; print "\$\$p-coverage,\$\$c"' tests/\$(TEST)`,\$(COVER_OPTIONS) tests/\$(TEST)

COVER_PARAMETERS = \$(PERL) -e '\$\$l = qx|grep __COVER__ \$\$ARGV[0]|; \$\$u = "-uncoverable_file \$\$1" if \$\$l =~ /__COVER__\\s+uncoverable_file\\s+(.*)/; (\$\$p) = \$\$l =~ /__COVER__\\s+cover_parameters\\s+(.*)/; print "\$\$u \$\$p"' tests/\$(TEST)

html : _run
\t \$(PERL) -Mblib cover `\$(COVER_PARAMETERS)` -report html

basic : _run
\t \$(PERL) -Mblib cover `\$(COVER_PARAMETERS)` -report html_basic

out : _run
\t \$(PERL) -Mblib cover `\$(COVER_PARAMETERS)` -report text > \$(TEST).out

text : out
\t \$(VISUAL) \$(TEST).out

wrun : pure_all
\t \$(PERL) \$(TAINT) -Iblib/lib -Iblib/arch -MDevel::Cover=-ignore,blib,-merge,0 tests/\$(TEST)

prove : pure_all
\t \$(PERL) -Iutils -MDevel::Cover::BuildUtils=prove_command -le '\$\$c = prove_command and print \$\$c and system \$\$c'

tst : pure_all
\t \$(PERL) -Iutils -MDevel::Cover::BuildUtils=prove_command -le '\$\$c = prove_command and print \$\$c and system \$\$c and system "make test"'

DB = cover_db

dump :
\t \$(PERL) -Mblib cover -dump_db \$(DB)

FONT = "Inconsolata 10"
GEOM = 260x85+0+0

diff : out
\t \$(PERL) utils/makeh strip_criterion 'time' \$(TEST).out
\t \$(PERL) utils/makeh strip_criterion ' pod' \$(TEST).out
\t gvim -geom \$(GEOM) -d -font \$(FONT) `\$(PERL) -Mblib -MDevel::Cover::Test -e '\$\$t = Devel::Cover::Test->new("\$(TEST)", run_test_at_end => 0); print \$\$t->cover_gold'` \$(TEST).out

gold : pure_all
\t \$(PERL) utils/create_gold \$(TEST)

all_test :
\t exec \$(PERL) utils/all_versions make test

all_gold :
\t rm -rf test_output
\t exec \$(PERL) utils/all_versions make gold

_delete_db : pure_all
\t rm -rf cover_db

_self_cover_tests : @{[sort values %tests]}
\t DEVEL_COVER_SELF=1 \$(PERL) -Mblib -MDevel::Cover cover -silent -write cover_db @{[sort values %tests]}

self_cover : _self_cover_reports
\t \$(PERL) -Mblib cover -report html_basic

mani :
\t -rm MANIFEST
\t \$(MAKE) manifest
\t \$(PERL) -i.bak -ne '\$\$d = "test_output/cover"; print unless /^\$\$d/; END { open M, ">>MANIFEST"; opendir D, \$\$d; print M "\$\$d/\$\$_\\n" for sort grep -f "\$\$d/\$\$_", readdir D }' MANIFEST

ok :
\t \@$Perlbug -okay  || echo "Please send your report manually to $Author"

nok :
\t \@$Perlbug -nokay || echo "Please send your report manually to $Author"
    ] . "\n" .

    join "\n",
         map("$tests{$_} : _delete_db\n" .
             "\t \@echo Running $tests{$_}\n" .
             "\t \@rm -rf $tests{$_}\n" .
             "\t \@DEVEL_COVER_SELF=1 \$(PERL) -Mblib -MDevel::Cover=-db,$tests{$_},-silent,1,-coverage,all,-ignore,tests/,-coverage,pod-also_private-xx $_\n",
             sort keys %tests),

         "_self_cover_reports : @{[map qq(report_$_), @reports]}\n",

         map("report_$_ : _self_cover_tests\n" .
             "\t \@echo Generating $_ report\n" .
             "\t \@DEVEL_COVER_SELF=1 \$(PERL) -Mblib -MDevel::Cover cover -silent -report $_ > /dev/null\n",
             @reports)
}
